In [48]:
import numpy as np
import pandas as pd
import statistics

In [49]:
# Question 1:
students = ['Adam', 'Anna', 'James', 'Corey', 'Alisha', 'Jodie', 'John']
grades = np.array([[11, 9, 16, 13.5],
                   [12, 11, 8.5, 10],
                   [11, 18, 11.5, 9],
                   [7, 15.5, 11, 14],
                   [9.5, 12, 10.5, 14],
                   [15, 18.5, 7, 12],
                   [18, 15.5, 11, 7.5]])

# Part 1
for i in range(len(students)):
    mean = np.mean(grades[i])
    print(f"{students[i]}'s grade average is {mean}")

# Part 2
math_grades = []
for i in range(len(students)):
    math_grades.append(grades[i][0])
median = np.median(math_grades)
print(f"The median math grade between all students is {median}.")

# Part 3
history_grades = []
for i in range(len(students)):
    history_grades.append(grades[i][3])
mode = statistics.mode(history_grades)
print(f"The most commonly occuring history grade between all students is {mode}.")
    

Adam's grade average is 12.375
Anna's grade average is 10.375
James's grade average is 12.375
Corey's grade average is 11.875
Alisha's grade average is 11.5
Jodie's grade average is 13.125
John's grade average is 13.0
The median math grade between all students is 11.0.
The most commonly occuring history grade between all students is 14.0.


In [50]:
# Question 2:
math, science, reading, history = [], [], [], []
for i in range(len(students)):
    math.append(grades[i][0])
    science.append(grades[i][1])
    reading.append(grades[i][2])
    history.append(grades[i][3])
df = pd.DataFrame({'Math': math,
                   'Science': science,
                   'Reading': reading,
                   'History': history})
df.corr()

# Math and History seem to have the strongest (albeit negative) correlation

Math   Science   Reading   History
Math     1.000000  0.269494 -0.268562 -0.718358
Science  0.269494  1.000000 -0.527606 -0.358688
Reading -0.268562 -0.527606  1.000000  0.206794
History -0.718358 -0.358688  0.206794  1.000000

In [51]:
# Question 3:
math_exam = [24, 5, 15, 60, 54, 82, 99, 80, 70, 98, 93, 60, 33, 22, 65, 61, 51, 58, 83, 86, 42, 67, 60]
def desc_stats(grade_list):
    print(f" mean: {np.mean(grade_list)}\n",
          f"median: {np.median(grade_list)}\n",
          f"mode: {statistics.mode(grade_list)}\n",
          f"minimum: {np.min(grade_list)}\n",
          f"maximum: {np.max(grade_list)}\n",
          f"range: {np.ptp(grade_list)}\n",
          f"variance: {np.var(grade_list)}\n",
          f"standard deviation: {np.std(grade_list)}\n",)

desc_stats(math_exam)

 mean: 59.47826086956522
 median: 60.0
 mode: 60
 minimum: 5
 maximum: 99
 range: 94
 variance: 664.1625708884687
 standard deviation: 25.771351747404882



Question 4:

In [61]:
curved_grades = []
for i in math_exam:
    curved_grades.append(i / 96.6)
sorted_grades = np.sort(math_exam)
print((np.sum(sorted_grades[-3:]) / 3))
print(np.mean(curved_grades))

96.66666666666667
0.6157169862273832


If I was the professor of this class I would curve the grade. How I would curve the grade would be how my history teacher used to curve grades. He took the average of the top 3 scores and used those as the 100% mark. So instead of 100 being 100, 96.6 would be the 100% mark. Although this wouldn't help much, it gives them a bit of a boost and that is all students need sometimes. Specifically, it would save 1 person (58) from failing and raise the class average from failing with 59.48% to passing with 61.57%. In summary, there is no downside to curving grades and it would help students a small amount.